design-proposals/kubevirt_ephemeral_csi_driver.md

```diff
@@ -0,0 +1,79 @@
+2. Something causes the virt-launcher to be force deleted (user action, resource pressure, whatever)
+3. When we normally terminate a VM, the virt-handler is notified we are going to terminate the VM, and is given a chance to unmount any volumes before the virt-launcher pod exits.
+4. When a pod that uses an emptyDir terminates, the kubelet will remove all the contents of the volume backing the emptyDir.
+5. Because we force terminated the virt-launcher pod, we are now in a race between the kubelet, and the virt-handler (normally the grace period will give virt-handler enough time). If the virt-handler is called to unmount and hotplugged volumes before the kubelet clears the emptyDir, we are fine. However if the kubelet is run before the virt-handler, it will blindly remove the contents of the emptyDir volume, which, in the hotplug volume case will be the disk.img files that have not been unmounted. Since it is a bind mount this will result in the disk.img file on the actual volumes being removed, and we have lost data.
+
+If we use a csi ephemeral volume controlled by a kubevirt specific csi driver, we can control the behavior in that scenario and unmount any mounted volumes inside the csi ephemeral volume before clearing it.
+
+## Goals
+
+* Design and implement a kubevirt specific csi driver that only supports **ephemeral volumes**.
+* Use csi ephemeral volumes to fully control what happens when virt-launcher pods terminate with regards to volume hotplug.
+
+## Non Goals
+
+Implement a full fledged CSI driver that also supports persistent volumes. We have other projects like the [hostpath csi driver](https://github.com/kubevirt/hostpath-provisioner) for that.
+
+
+## User Stories
+
+* As a KubeVirt user I want to be guaranteed that if I have hotplugged volumes in my VM, and the virt-launcher pod terminates unexpectedly, I will not lose any data on my volumes.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The idea is to implement just the ephemeral parts of the csi spec. Examples of drivers that do this are [csi-driver-image-populate](https://github.com/kubernetes-csi/csi-driver-image-populator) and [secrets-store-csi-driver](https://github.com/kubernetes-sigs/secrets-store-csi-driver).
+
+To support just the ephemeral portions, we need to implement the appropriate parts of the csi driver, and start the [node-driver-registrar](https://github.com/kubernetes-csi/node-driver-registrar) so we can register the driver with kubernetes. The csi driver will have to run on each node, so will require a daemonset. We could make an additional daemonset, or add additional containers to the virt-handler daemonset. The daemonset will require privileged permissions due to the nature of gRPC socket used to communicate between the different CSI components.
```

Comment:
as an initial PoC, starting with building this logic into virt-handler sounds like a good plan. We already have device plugins built into virt-handler as well, so there's a precedence for this kind of thing.

Reply:
I was just pointing out that we would have some extra containers in virt-handler. In particular the driver registrar and the actual csi driver which are two separate containers.